contracts/account/extensions/AccountERC7579.sol

```diff
@@ -0,0 +1,370 @@
+     * rest of the data to be sent to the handler when calling {IERC7579Module-onUninstall}.
+     *
+     * Requirements:
+     *
+     * * Module must be already installed. Reverts with {ERC7579UninstalledModule} otherwise.
+     */
+    function _uninstallModule(uint256 moduleTypeId, address module, bytes memory deInitData) internal virtual {
+        require(
+            (moduleTypeId != MODULE_TYPE_VALIDATOR || _validators.remove(module)) &&
+                (moduleTypeId != MODULE_TYPE_EXECUTOR || _executors.remove(module)),
+            ERC7579Utils.ERC7579UninstalledModule(moduleTypeId, module)
+        );
+
+        if (moduleTypeId == MODULE_TYPE_FALLBACK) {
+            bytes4 selector;
+            (selector, deInitData) = abi.decode(deInitData, (bytes4, bytes));
+            require(_uninstallFallback(module, selector), ERC7579Utils.ERC7579UninstalledModule(moduleTypeId, module));
+        }
+
+        IERC7579Module(module).onUninstall(deInitData);
+        emit ModuleUninstalled(moduleTypeId, module);
+    }
+
+    /**
+     * @dev Installs a fallback handler for the given selector. Returns true if the handler was installed,
+     */
+    function _installFallback(address module, bytes4 selector) internal virtual returns (bool) {
+        if (_fallbacks[selector] != address(0)) return false;
+        _fallbacks[selector] = module;
+        return true;
+    }
+
+    /**
+     * @dev Uninstalls a fallback handler for the given selector. Returns true if the handler was uninstalled,
+     */
+    function _uninstallFallback(address module, bytes4 selector) internal virtual returns (bool) {
+        address handler = _fallbackHandler(selector);
+        if (handler == address(0) || handler != module) return false;
+        delete _fallbacks[selector];
+        return true;
+    }
+
+    /// @dev Returns the fallback handler for the given selector. Returns `address(0)` if not installed.
+    function _fallbackHandler(bytes4 selector) internal view virtual returns (address) {
+        return _fallbacks[selector];
+    }
+
+    /// @dev Extracts the nonce validator from the user operation.
+    ///
+    /// To construct a nonce key, set nonce as follows:
+    ///
+    /// ```
+    /// <module address (20 bytes)> | <key (4 bytes)> | <nonce (8 bytes)>
+    /// ```
+    function _extractUserOpValidator(PackedUserOperation calldata userOp) internal pure virtual returns (address) {
+        return address(bytes32(userOp.nonce).extract_32_20(0));
+    }
+
+    /// @dev Extracts the signature validator from the signature.
+    ///
+    /// To construct a signature, set the first 20 bytes as the module address and the remaining bytes as the
+    /// signature data:
+    ///
+    /// ```
+    /// <module address (20 bytes)> | <signature data>
+    /// ```
+    function _extractSignatureValidator(bytes calldata signature) internal pure virtual returns (address) {
+        return address(bytes20(signature[0:20]));
+    }
```

Comment:
I think having two different mechanisms for extracting the validator based on the usecase is really confusing. IMO we should only have one.

Reply:
I just realised that removing the `_validateUserOp` override and relying on `AccountCore._validateUserOp` calling `_rawSignatureValidation` could work, but it would use `IERC7579Validator.isValidSignatureWithSender` instead of `IERC7579Validator.validateUserOp`... what a mess

using the nonce is a nice trick because it avoid copying the entier userOp to memory to rewrite the signature before sending is again to the module ... but still it feels very hacky (and inconsistent in the way signatures are generated)

Reply:
If there was a standard encoding of the module address in the signature, and if the module knew about it, it would be cleaner.

Not sure if we can assume the module know it will have to drop the first 20 bytes of the userOp signature that are used for module selection. Maybe we should try to make that "standard validation module behavior"